sourcegraph/cody
# Edit: Track rejection rate

## Description

This PR adds an rejection metric that will listen for the next action in a document immediately after an Edit.

It works like this:
- We track how often the user hits "Undo" or "Redo" immediately after an Edit.
- When a user makes a new insertion to a document, we consider them to have moved on from the Edit.
- If we observed that the user undid the Edit, then we mark it as "rejected", otherwise we mark it as "accepted"
- If the user clicks the "Undo" command from the Edit codelens, we consider this to be a rejection, and immediately mark it as such.

The intention here is to provide a secondary metric to our persistence rate. We cannot track acceptance rate, as edits are automatically inserted to the document, so we essentially need to do the inverse, and check if the user rejects the change or not. There is a chance that this is not 100% accurate, as a user may manually modify the edit output if they are not happy with it. However this could be seen as both a partial acceptance or partial rejection, so the persistence rate is likely more useful here.

## Test plan

Undo:

1. Make an edit
2. Undo it
3. Make another change
4. Check event fired as `fixup:user:rejected`

Undo/Redo

1. Make an Edit
2. Undo it
3. Redo to apply the change again
4. Make another change
5. Check event fired as `fixup:user:accepted`

Undo via Code lens

1. Make an Edit
2. Undo it
3. Redo to apply the change again
4. Make another change
5. Check event fired as `fixup:user:rejected`

Happy path (any other change):
1. Make an edit
2. Make another change
3. Check event fired as `fixup:user:accepted`
